### PR TITLE
#40 Specifications are not checked for overlapping input and output action sets

### LINKED ISSUE #40: Specifications are not checked for overlapping input and output action sets
There seems to be no precheck whether the action set of a specification is a partition (i.e. whether each action is either an input *xor* an output action).

### PR DESCRIPTION
**Pull request by Brandhoej**
The Automaton constructor checks whether an action is either an input xor an output.

# Description
* The main Automaton constructor now checks whether an action is either an input xor output. This is only done in the Automaton constructor (per definition 2) and not in the TransitionSystem (per definition 1), as the TransitionSystem constructor requires an Automaton, and it thereby inherit this property.
* Automaton checks for a minimum of one location in its constructor now.
* Added for tests for the Automaton construction.
* Added a test for Automaton copy constructor
* Fixed the copy constructor for Automaton which did not create new instances of the action sets
* Made addTargetInvariantToEdges and makeInputEnabled private but I expect it to be a part of the public API later. I think for it to be a part of the public API maybe the addTargetInvariantToEdges should be a part of makeInputEnabled.
* Made the action sets finals as a completely new instance of these sets should not be expected throughout the lifetime of an Automaton.

# Related issue
Closes #40 

# How Has This Been Tested?
The added tests for the Automaton, and no test fails because of this change.

### CODE REVIEW

File: src/models/Automaton.java [hunk 1]
```diff
@@ -17,6 +17,10 @@ public class Automaton {
     private final Location initial;
 
     public Automaton(String name, List<Location> locations, List<Edge> edges, List<Clock> clocks, List<BoolVar> BVs, boolean makeInputEnabled) {
+        if (locations.isEmpty()) {
+            throw new IllegalArgumentException("An automaton must have atleast one location");
```
Comment (magoorden) [inline src/models/Automaton.java:21]:
```suggestion
            throw new IllegalArgumentException("An automaton must have at least one location.");
```

Or even

```suggestion
            throw new IllegalArgumentException(String.format("Automaton %s has no location.", name));
```

File: src/models/Automaton.java [hunk 2]
```diff
@@ -46,6 +51,23 @@ public Automaton(String name, List<Location> locations, List<Edge> edges, List<C
...
+        /* The finite set of actions must be partitioned into inputs and outputs.
+         *   Here we check whether they are partitioned by checking that the intersection of the
+         *   inputs and outputs are empty as an action can only be either an input xor an output.
+         *   We don't have to check whether an action is neither an input nor an output as the set
+         *   of actions is build as the union of the inputs and outputs and for this reason
+         *   guarantees to be in the set of actions. */
+        Set<Channel> intersection = new HashSet<>(inputAct);
+        intersection.retainAll(outputAct);
+        if (!intersection.isEmpty()) {
+            throw new IllegalArgumentException("The action set of a specification must be a partition, where each action is either an input xor an output");
```
Comment (magoorden) [inline src/models/Automaton.java:63]:
```suggestion
            throw new IllegalArgumentException("The action set of a specification must be a partition, where each action is either an input xor an output.");
```

Or more indicative for the end user where the problem is.

```suggestion
            throw new IllegalArgumentException(String.format("The action set of specification automaton %s is not a partition.", name));
```
You could even extend this by finding a human readable way of printing all the elements of `intersection`, which could be just a single element or a lot of elements.

File: src/models/Automaton.java [hunk 3]
```diff
@@ -46,6 +51,23 @@ public Automaton(String name, List<Location> locations, List<Edge> edges, List<C
...
+        }
+
+        /* As the inputs and outputs are shown to be disjoint then the union of them
+         *   it holds that the disjoint union of inputs and outputs is a finite set
+         *   of actions partitioned into inputs and outputs. Here the "disjoint" comes
+         *   from actions (or Channels) are reference types and are for this reason not
+         *   merged with actions of the same name. */
```
Comment (magoorden) [inline src/models/Automaton.java:70]: Comment has some English grammar issues, but I mainly think it's too wordy for what you actually try to convey.

Maybe something like

`Since inputAct and outputAct are now disjoint, we can simply construct actions as the union of both sets.`

**Review by magoorden — COMMENTED** [review]
I found just some small things. 

I like that you also created test cases for no and multiple initial locations.

File: src/models/Automaton.java (hunk 1 above)
Comment (Brandhoej) [reply src/models/Automaton.java]: Good idea. I have changed it to ```String.format("Automaton %s cannot have more than one initial location", name)```

File: src/models/Automaton.java (hunk 2 above)
Comment (Brandhoej) [reply src/models/Automaton.java]: I have changed it to ```throw new IllegalArgumentException(String.format("The actions %s of specification automaton %s is not a partition.", violatingActions, name));``` where ```violatingActions``` is a string joined by the names of all actions in the intersection "{a, b, c}". Is the message in the exception still correct? Maybe we should also consider encapsulating the cause as a Throwable for the IllegalArgument Exception? This could be done by creating a ```Throwable``` ```PartitionException``` and its constructor might take all actions or just the violating ones, then an instance of the ```PartitionException``` could be used as the cause of the ```IllegalArgumentException```. Overall I think the creation of a ```PartitionException``` in this case is a good idea because the ```violatingActions``` might be a very large string.

File: src/models/Automaton.java (hunk 3 above)
Comment (Brandhoej) [reply src/models/Automaton.java]: Very good and comment.

File: src/models/Automaton.java [hunk 4]
```diff
@@ -269,18 +281,11 @@ public void makeInputEnabled() {
...
-        for (Edge edge : getEdges()) {
-            CDD targetCDD = edge.getTarget().getInvariantCDD();
-            CDD past = targetCDD.transitionBack(edge);
-            if (!past.equiv(CDD.cddTrue()))
-                edge.setGuard(past.conjunction(edge.getGuardCDD()).getGuard(getClocks()));
+        if (initialisedCdd) {
+            CDD.done();
         }
     }
 }
```
Comment (magoorden) [inline src/models/Automaton.java:291]: I see that Github is 'complaining' that this file does not have a new line character at the end 😆 .

File: src/models/Automaton.java (hunk 2 above)
Comment (magoorden) [reply src/models/Automaton.java]: Technically the exception text `"The actions %s of specification automaton %s is not a partition."`, because partition is a property of the complete action set. But we only want to display the actions that are causing the action set to be *not* a partition. 
Since this list of actions can vary in size (and potentiallly be very long), maybe it is more user friendly to start with the fixed length part like `"The actions set of specification automaton %s is not a partition, caused by %s."`. When we create a `PartitionException`, we can let that exception do the formatting. So it will also be easier to change if we want to.

**Comment by magoorden** [comment]
Good corrections :D

**Comment by magoorden** [comment]
Some other think I thought about just before lunch (but I think it's a little bit out of scope of this pull request) is to collect all the problems before terminating the program. Right now an `IllegalArgumentException` is thrown the moment the first problem is encountered, but there might be many problems in the model.
This is something similar to the semantic check in Uppaal, where it lists all the problems.

**Comment by Brandhoej** [comment]
> Some other think I thought about just before lunch (but I think it's a little bit out of scope of this pull request) is to collect all the problems before terminating the program. Right now an `IllegalArgumentException` is thrown the moment the first problem is encountered, but there might be many problems in the model. This is something similar to the semantic check in Uppaal, where it lists all the problems.

I believe this is a very good idea. Aggregating exceptions can be very hard as an incorrect result which originally would have exited the function would then propagate down the function and cause other exceptions. This might make it hard to understand which exception is the root cause. From my understanding it would be idiomatic to create separate classes focusing on the various validation aspects, such as disjoint input/outputs, and then allow it to throw exceptions which are then wrapped with [ValidationException](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/ValidationException.html). Moving this validation behavior out of the Automaton constructor would also allow us to give user feedback without  actually constructing an Automaton, which might execute some behavior we dont care about. 

I suggest that we open a new issue which focuses on experimenting with the validation pattern and its usage for validating the constructor for the Automaton, and then later we could apply it in other areas such as the Composition/Conjunction.

**Comment by magoorden** [comment]
This PR is still labeled as Draft. Are you planning on including more stuff?
